File: app/src/main/java/org/skyfaced/mvp/util/WaifuType.java
package org.skyfaced.mvp.util;

public class WaifuType {
    public String getName() {
        throw new IllegalArgumentException("Override this method.");
    }

    public String getCategory() {
        throw new IllegalArgumentException("Override this method.");
    }

    public static class SFW extends WaifuType {
        private final Category category;

        public SFW(Category category) {
            this.category = category;
        }

        @Override
        public String getName() {
            return this.getClass().getSimpleName().toLowerCase();
        }

        @Override
        public String getCategory() {
            return category.name().toLowerCase();
        }

        public enum Category {
            Waifu,
            Neko,
            Shinobu,
            Megumin,
            Bully,
            Cuddle,
            Cry,
            Hug,
            Awoo,
            Kiss,
            Lick,
            Pat,
            Smug,
            Bonk,
            Yeet,
            Blush,
            Smile,
            Wave,
            Highfive,
            Handhold,
            Nom,
            Bite,
            Glomp,
            Slap,
            Kill,
            Happy,
            Wink,
            Poke,
            Dance,
            Cringe
        }
    }

    public static class NSFW extends WaifuType {
        private final Category category;

        public NSFW(Category category) {
            this.category = category;
        }

        @Override
        public String getName() {
            return this.getClass().getSimpleName().toLowerCase();
        }

        @Override
        public String getCategory() {
            return category.name().toLowerCase();
        }

        public enum Category {
            Waifu,
            Neko,
            Trap,
            Blowjob
        }
    }
}

File: app/src/main/java/org/skyfaced/mvp/util/WaifuType.kt
package org.skyfaced.mvp.util

sealed class WaifuType(open val innerCategory: Category) {
    class SFW(override val innerCategory: InnerCategory) : WaifuType(innerCategory) {
        enum class InnerCategory : Category {
            Waifu,
            Neko,
            Shinobu,
            Megumin,
            Bully,
            Cuddle,
            Cry,
            Hug,
            Awoo,
            Kiss,
            Lick,
            Pat,
            Smug,
            Bonk,
            Yeet,
            Blush,
            Smile,
            Wave,
            Highfive,
            Handhold,
            Nom,
            Bite,
            Glomp,
            Slap,
            Kill,
            Happy,
            Wink,
            Poke,
            Dance,
            Cringe;
        }

        override val type: String
            get() = this::class.java.simpleName.lowercase()

        override val category: String
            get() = innerCategory.name.lowercase()
    }

    class NSFW(override val innerCategory: InnerCategory) : WaifuType(innerCategory) {
        enum class InnerCategory : Category {
            Waifu,
            Neko,
            Trap,
            Blowjob;
        }

        override val type: String
            get() = this::class.java.simpleName.lowercase()

        override val category: String
            get() = innerCategory.name.lowercase()
    }

    abstract val type: String

    abstract val category: String
}

interface Category
